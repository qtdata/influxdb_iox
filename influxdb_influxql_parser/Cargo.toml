[package]
name = "influxdb_influxql_parser"
version.workspace = true
authors.workspace = true
edition.workspace = true
license.workspace = true

[dependencies] # In alphabetical order
nom = { version = "7", default-features = false, features = ["std"] }
once_cell = "1"
workspace-hack = { path = "../workspace-hack"}

[dev-dependencies] # In alphabetical order
test_helpers = { path = "../test_helpers" }
assert_matches = "1"
insta = { version = "1.23.0", features = ["yaml"] }