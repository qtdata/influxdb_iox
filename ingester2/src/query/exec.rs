use async_trait::async_trait;
use data_types::{NamespaceId, TableId};
use observability_deps::tracing::*;
use trace::span::{Span, SpanRecorder};

use super::{QueryError, QueryExec};
use crate::query::response::QueryResponse;

#[derive(Debug)]
pub(crate) struct QueryRunner;

impl QueryRunner {
    pub(crate) fn new() -> Self {
        Self {}
    }
}

#[async_trait]
impl QueryExec for QueryRunner {
    type Response = QueryResponse;

    async fn query_exec(
        &self,
        namespace_id: NamespaceId,
        table_id: TableId,
        columns: Vec<String>,
        span: Option<Span>,
    ) -> Result<Self::Response, QueryError> {
        let mut _span_recorder = SpanRecorder::new(span);

        info!(
            namespace_id=%namespace_id,
            table_id=%table_id,
            columns=?columns,
            "executing query"
        );

        unimplemented!();
    }
}
