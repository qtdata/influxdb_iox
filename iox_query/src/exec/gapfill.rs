//! This module contains code that implements
//! a gap-filling extension to DataFusion

use std::{
    fmt::{self, Debug},
    sync::Arc,
};

use arrow::{compute::SortOptions, datatypes::SchemaRef};
use datafusion::{
    common::DFSchemaRef,
    error::{DataFusionError, Result},
    execution::context::TaskContext,
    logical_expr::{LogicalPlan, UserDefinedLogicalNode},
    physical_expr::{create_physical_expr, execution_props::ExecutionProps, PhysicalSortExpr},
    physical_plan::{
        expressions::Column, DisplayFormatType, Distribution, ExecutionPlan, Partitioning,
        PhysicalExpr, SendableRecordBatchStream, Statistics,
    },
    prelude::Expr,
};

/// A logical node that represents the gap filling operation.
#[derive(Clone, Debug)]
pub struct GapFill {
    input: Arc<LogicalPlan>,
    group_expr: Vec<Expr>,
    aggr_expr: Vec<Expr>,
    time_column: Expr,
    stride: Expr,
}

impl GapFill {
    pub fn try_new(
        input: Arc<LogicalPlan>,
        group_expr: Vec<Expr>,
        aggr_expr: Vec<Expr>,
        time_column: Expr,
        stride: Expr,
    ) -> Result<Self> {
        Ok(Self {
            input,
            group_expr,
            aggr_expr,
            time_column,
            stride,
        })
    }
}

impl UserDefinedLogicalNode for GapFill {
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    fn inputs(&self) -> Vec<&LogicalPlan> {
        vec![self.input.as_ref()]
    }

    fn schema(&self) -> &DFSchemaRef {
        self.input.schema()
    }

    fn expressions(&self) -> Vec<Expr> {
        self.group_expr
            .iter()
            .chain(self.aggr_expr.iter())
            .cloned()
            .collect()
    }

    fn fmt_for_explain(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "GapFill: groupBy=[{:?}], aggr=[{:?}], time_column={}, stride={}",
            self.group_expr, self.aggr_expr, self.time_column, self.stride
        )
    }

    fn from_template(
        &self,
        exprs: &[Expr],
        inputs: &[LogicalPlan],
    ) -> Arc<dyn UserDefinedLogicalNode> {
        let mut group_expr: Vec<_> = exprs.to_vec();
        let aggr_expr = group_expr.split_off(self.group_expr.len());
        let gapfill = Self::try_new(
            Arc::new(inputs[0].clone()),
            group_expr,
            aggr_expr,
            self.time_column.clone(),
            self.stride.clone(),
        )
        .expect("should not fail");
        Arc::new(gapfill)
    }
}

/// Called by the extension planner to plan a [GapFill] node.
pub(crate) fn plan_gap_fill(
    execution_props: &ExecutionProps,
    gap_fill: &GapFill,
    logical_inputs: &[&LogicalPlan],
    physical_inputs: &[Arc<dyn ExecutionPlan>],
) -> Result<GapFillExec> {
    if logical_inputs.len() != 1 {
        return Err(DataFusionError::Internal(
            "GapFillExec: wrong number of logical inputs".to_string(),
        ));
    }
    if physical_inputs.len() != 1 {
        return Err(DataFusionError::Internal(
            "GapFillExec: wrong number of physical inputs".to_string(),
        ));
    }

    let input_dfschema = logical_inputs[0].schema().as_ref();
    let input_schema = physical_inputs[0].schema();
    let input_schema = input_schema.as_ref();

    let group_expr: Result<Vec<_>> = gap_fill
        .group_expr
        .iter()
        .map(|e| create_physical_expr(e, input_dfschema, input_schema, execution_props))
        .collect();
    let group_expr = group_expr?;

    let aggr_expr: Result<Vec<_>> = gap_fill
        .aggr_expr
        .iter()
        .map(|e| create_physical_expr(e, input_dfschema, input_schema, execution_props))
        .collect();
    let aggr_expr = aggr_expr?;

    let logical_time_column = gap_fill.time_column.try_into_col()?;
    let time_column = Column::new_with_schema(&logical_time_column.name, input_schema)?;

    let stride = create_physical_expr(
        &gap_fill.stride,
        input_dfschema,
        input_schema,
        execution_props,
    )?;

    GapFillExec::try_new(
        Arc::clone(&physical_inputs[0]),
        group_expr,
        aggr_expr,
        time_column,
        stride,
    )
}

/// A physical node for the gap-fill operation.
pub struct GapFillExec {
    input: Arc<dyn ExecutionPlan>,
    // The group by expressions from the original aggregation node.
    group_expr: Vec<Arc<dyn PhysicalExpr>>,
    // The aggregate expressions from the original aggregation node.
    aggr_expr: Vec<Arc<dyn PhysicalExpr>>,
    // The column of binned times to be gap-filled.
    time_column: Column,
    // The sort expressions for the required sort order of the input:
    // all of the group exressions, with the time column being last.
    sort_expr: Vec<PhysicalSortExpr>,
    // The stride of the date bins
    stride: Arc<dyn PhysicalExpr>,
}

impl GapFillExec {
    pub fn try_new(
        input: Arc<dyn ExecutionPlan>,
        group_expr: Vec<Arc<dyn PhysicalExpr>>,
        aggr_expr: Vec<Arc<dyn PhysicalExpr>>,
        time_column: Column,
        stride: Arc<dyn PhysicalExpr>,
    ) -> Result<Self> {
        let sort_expr = {
            let mut sort_expr: Vec<_> = group_expr
                .iter()
                .map(|expr| PhysicalSortExpr {
                    expr: Arc::clone(expr),
                    options: SortOptions::default(),
                })
                .collect();

            // Ensure that the time column is the last component in the sort
            // expressions.
            let time_idx = group_expr
                .iter()
                .enumerate()
                .find(|(_i, e)| {
                    if let Some(col) = e.as_any().downcast_ref::<Column>() {
                        col.index() == time_column.index()
                    } else {
                        false
                    }
                })
                .map(|(i, _)| i);

            if let Some(time_idx) = time_idx {
                let last_elem = sort_expr.len() - 1;
                sort_expr.swap(time_idx, last_elem);
            } else {
                return Err(DataFusionError::Internal(
                    "could not find time column for GapFillExec".to_string(),
                ));
            }

            sort_expr
        };

        Ok(Self {
            input,
            group_expr,
            aggr_expr,
            time_column,
            sort_expr,
            stride,
        })
    }
}

impl Debug for GapFillExec {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "GapFillExec")
    }
}

impl ExecutionPlan for GapFillExec {
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    fn schema(&self) -> SchemaRef {
        self.input.schema()
    }

    fn output_partitioning(&self) -> Partitioning {
        Partitioning::UnknownPartitioning(1)
    }

    fn required_input_distribution(&self) -> Vec<Distribution> {
        // It seems like it could be possible to partition on all the
        // group keys except for the time expression. For now, keep it simple.
        vec![Distribution::SinglePartition]
    }

    fn output_ordering(&self) -> Option<&[datafusion::physical_expr::PhysicalSortExpr]> {
        self.input.output_ordering()
    }

    fn required_input_ordering(&self) -> Vec<Option<&[PhysicalSortExpr]>> {
        vec![Some(&self.sort_expr)]
    }

    fn maintains_input_order(&self) -> Vec<bool> {
        vec![true]
    }

    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
        vec![Arc::clone(&self.input)]
    }

    fn with_new_children(
        self: Arc<Self>,
        children: Vec<Arc<dyn ExecutionPlan>>,
    ) -> Result<Arc<dyn ExecutionPlan>> {
        match children.len() {
            1 => Ok(Arc::new(Self::try_new(
                Arc::clone(&children[0]),
                self.group_expr.clone(),
                self.aggr_expr.clone(),
                self.time_column.clone(),
                Arc::clone(&self.stride),
            )?)),
            _ => Err(DataFusionError::Internal(
                "GapFillExec wrong number of children".to_string(),
            )),
        }
    }

    fn execute(
        &self,
        partition: usize,
        _context: Arc<TaskContext>,
    ) -> Result<SendableRecordBatchStream> {
        if self.output_partitioning().partition_count() <= partition {
            return Err(DataFusionError::Internal(format!(
                "GapFillExec invalid partition {partition}"
            )));
        }
        Err(DataFusionError::NotImplemented("gap filling".to_string()))
    }

    fn fmt_as(&self, t: DisplayFormatType, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match t {
            DisplayFormatType::Default => {
                let group_expr: Vec<_> = self.group_expr.iter().map(|e| e.to_string()).collect();
                let aggr_expr: Vec<_> = self.aggr_expr.iter().map(|e| e.to_string()).collect();
                write!(
                    f,
                    "GapFillExec: group_expr=[{}], aggr_expr=[{}], stride={}",
                    group_expr.join(", "),
                    aggr_expr.join(", "),
                    self.stride,
                )
            }
        }
    }

    fn statistics(&self) -> Statistics {
        Statistics::default()
    }
}

#[cfg(test)]
mod test {
    use crate::exec::{Executor, ExecutorType};

    use super::*;
    use arrow::datatypes::{DataType, Field, Schema, TimeUnit};
    use datafusion::{
        datasource::empty::EmptyTable,
        error::Result,
        logical_expr::{logical_plan, Extension},
        physical_plan::displayable,
        prelude::{col, lit},
        scalar::ScalarValue,
        sql::TableReference,
    };

    fn schema() -> Schema {
        Schema::new(vec![
            Field::new(
                "time",
                DataType::Timestamp(TimeUnit::Nanosecond, None),
                false,
            ),
            Field::new("loc", DataType::Utf8, false),
            Field::new("temp", DataType::Float64, false),
        ])
    }

    fn table_scan() -> Result<LogicalPlan> {
        let schema = schema();
        logical_plan::table_scan(Some("temps"), &schema, None)?.build()
    }

    #[test]
    fn fmt_logical_plan() -> Result<()> {
        // This test case does not make much sense but
        // just verifies we can construct a logical gapfill node
        // and show its plan.
        let scan = table_scan()?;
        let gapfill = GapFill::try_new(
            Arc::new(scan),
            vec![col("loc"), col("time")],
            vec![col("temp")],
            col("time"),
            lit(ScalarValue::IntervalDayTime(Some(60_000))),
        )?;
        let plan = LogicalPlan::Extension(Extension {
            node: Arc::new(gapfill),
        });
        let expected = "GapFill: groupBy=[[loc, time]], aggr=[[temp]], time_column=time, stride=IntervalDayTime(\"60000\")\
                      \n  TableScan: temps";
        assert_eq!(expected, format!("{}", plan.display_indent()));
        Ok(())
    }

    async fn assert_explain(sql: &str, expected: &str) -> Result<()> {
        let executor = Executor::new_testing();
        let context = executor.new_context(ExecutorType::Query);
        context.inner().register_table(
            TableReference::Bare { table: "temps" },
            Arc::new(EmptyTable::new(Arc::new(schema()))),
        )?;
        let physical_plan = context.prepare_sql(sql).await?;
        let actual_plan = displayable(physical_plan.as_ref()).indent().to_string();
        let actual_iter = actual_plan.split('\n');

        let expected = expected.split('\n');
        expected.zip(actual_iter).for_each(|(expected, actual)| {
            assert_eq!(
                expected, actual,
                "\ncomplete plan was:\n{:?}\n",
                actual_plan
            )
        });
        Ok(())
    }

    #[tokio::test]
    async fn plan_gap_fill() -> Result<()> {
        // show that the optimizer rule can fire and that physical
        // planning will succeed.
        let dbg_args = "IntervalDayTime(\"60000\"),temps.time,Utf8(\"1970-01-01T00:00:00Z\")";
        assert_explain(
            "SELECT date_bin_gapfill(interval '1 minute', time, timestamp '1970-01-01T00:00:00Z') AS minute, avg(temp)\
           \nFROM temps\
           \nGROUP BY minute;",
            format!(
                "ProjectionExec: expr=[date_bin_gapfill({dbg_args})@0 as minute, AVG(temps.temp)@1 as AVG(temps.temp)]\
               \n  GapFillExec: group_expr=[date_bin_gapfill({dbg_args})@0], aggr_expr=[AVG(temps.temp)@1], stride=60000\
               \n    SortExec: [date_bin_gapfill({dbg_args})@0 ASC]\
               \n      AggregateExec: mode=Final, gby=[date_bin_gapfill({dbg_args})@0 as date_bin_gapfill({dbg_args})], aggr=[AVG(temps.temp)]"
           ).as_str()
       ).await?;
        Ok(())
    }

    #[tokio::test]
    async fn gap_fill_exec_sort_order() -> Result<()> {
        // The call to `date_bin_gapfill` should be last in the SortExec
        // expressions, even though it was not last on the SELECT list
        // or the GROUP BY clause.
        let dbg_args = "IntervalDayTime(\"60000\"),temps.time,Utf8(\"1970-01-01T00:00:00Z\")";
        assert_explain(
            "SELECT \
           \n  loc,\
           \n  date_bin_gapfill(interval '1 minute', time, timestamp '1970-01-01T00:00:00Z') AS minute,\
           \n  concat('zz', loc) AS loczz,\
           \n  avg(temp)\
           \nFROM temps\
           \nGROUP BY loc, minute, loczz;",
            format!(
                "ProjectionExec: expr=[loc@0 as loc, date_bin_gapfill({dbg_args})@1 as minute, concat(Utf8(\"zz\"),temps.loc)@2 as loczz, AVG(temps.temp)@3 as AVG(temps.temp)]\
               \n  GapFillExec: group_expr=[loc@0, date_bin_gapfill({dbg_args})@1, concat(Utf8(\"zz\"),temps.loc)@2], aggr_expr=[AVG(temps.temp)@3], stride=60000\
               \n    SortExec: [loc@0 ASC,concat(Utf8(\"zz\"),temps.loc)@2 ASC,date_bin_gapfill({dbg_args})@1 ASC]\
               \n      AggregateExec: mode=Final, gby=[loc@0 as loc, date_bin_gapfill({dbg_args})@1 as date_bin_gapfill({dbg_args}), concat(Utf8(\"zz\"),temps.loc)@2 as concat(Utf8(\"zz\"),temps.loc)], aggr=[AVG(temps.temp)]"
           ).as_str()

           ).await?;
        Ok(())
    }
}
