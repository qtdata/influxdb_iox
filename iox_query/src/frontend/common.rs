use std::sync::Arc;

use datafusion::{
    catalog::TableReference, common::tree_node::TreeNode, datasource::provider_as_source,
    logical_expr::LogicalPlanBuilder,
};
use observability_deps::tracing::trace;
use predicate::Predicate;
use schema::Schema;
use snafu::{ResultExt, Snafu};

use crate::{
    provider::{ChunkTableProvider, ProviderBuilder},
    util::MissingColumnsToNull,
    QueryChunk,
};

#[derive(Debug, Snafu)]
pub enum Error {
    #[snafu(display(
        "gRPC planner got error adding chunk for table {}: {}",
        table_name,
        source
    ))]
    CreatingProvider {
        table_name: String,
        source: crate::provider::Error,
    },

    #[snafu(display(
        "Internal gRPC planner rewriting predicate for {}: {}",
        table_name,
        source
    ))]
    RewritingFilterPredicate {
        table_name: String,
        source: datafusion::error::DataFusionError,
    },

    #[snafu(display("Error building plan: {}", source))]
    BuildingPlan {
        source: datafusion::error::DataFusionError,
    },
}

pub(crate) type Result<T, E = Error> = std::result::Result<T, E>;

/// Represents scanning one or more [`QueryChunk`]s.
pub struct ScanPlan {
    pub plan_builder: LogicalPlanBuilder,
    pub provider: Arc<ChunkTableProvider>,
}

impl std::fmt::Debug for ScanPlan {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ScanPlan")
            .field("plan_builder", &"<...>")
            .field("provider", &self.provider)
            .finish()
    }
}

impl ScanPlan {
    /// Return the schema of the source (the merged schema across all tables)
    pub fn schema(&self) -> &Schema {
        self.provider.iox_schema()
    }
}

/// Builder for [`ScanPlan`]s which scan the data  1 or more [`QueryChunk`] for
/// IOx's custom query frontends (InfluxRPC and Reorg at the time of
/// writing).
///
/// The created plan looks like:
///
/// ```text
///   Filter(predicate) [optional]
///     Scan
/// ```
///
/// NOTE: This function assumes the chunks have already been "pruned"
/// based on statistics and will not attempt to prune them
/// further. Some frontends like influxrpc or the reorg planner manage
/// (and thus prune) their own chunklist.

#[derive(Debug)]
pub struct ScanPlanBuilder<'a> {
    table_name: Arc<str>,
    /// The schema of the resulting table (any chunks that don't have
    /// all the necessary columns will be extended appropriately)
    table_schema: &'a Schema,
    chunks: Vec<Arc<dyn QueryChunk>>,
    predicate: Option<&'a Predicate>,
    /// Do deduplication
    deduplication: bool,
}

impl<'a> ScanPlanBuilder<'a> {
    pub fn new(table_name: Arc<str>, table_schema: &'a Schema) -> Self {
        Self {
            table_name,
            table_schema,
            chunks: vec![],
            predicate: None,
            // always do deduplication in query
            deduplication: true,
        }
    }

    /// Adds `chunks` to the list of Chunks to scan
    pub fn with_chunks(mut self, chunks: impl IntoIterator<Item = Arc<dyn QueryChunk>>) -> Self {
        self.chunks.extend(chunks.into_iter());
        self
    }

    /// Sets the predicate
    pub fn with_predicate(mut self, predicate: &'a Predicate) -> Self {
        assert!(self.predicate.is_none());
        self.predicate = Some(predicate);
        self
    }

    /// Deduplication
    pub fn enable_deduplication(mut self, deduplication: bool) -> Self {
        self.deduplication = deduplication;
        self
    }

    /// Creates a `ScanPlan` from the specified chunks
    pub fn build(self) -> Result<ScanPlan> {
        let Self {
            table_name,
            chunks,
            table_schema,
            predicate,
            deduplication,
        } = self;

        assert!(!chunks.is_empty(), "no chunks provided");

        // Prepare the plan for the table
        let mut builder = ProviderBuilder::new(Arc::clone(&table_name), table_schema.clone())
            .with_enable_deduplication(deduplication);

        for chunk in chunks {
            builder = builder.add_chunk(chunk);
        }

        let provider = builder.build().context(CreatingProviderSnafu {
            table_name: table_name.as_ref(),
        })?;

        let provider = Arc::new(provider);
        let source = provider_as_source(Arc::clone(&provider) as _);

        // Scan all columns (DataFusion optimizer will prune this
        // later if possible)
        let projection = None;

        // Do not parse the tablename as a SQL identifer, but use as is
        let table_ref = TableReference::bare(table_name.to_string());
        let mut plan_builder =
            LogicalPlanBuilder::scan(table_ref, source, projection).context(BuildingPlanSnafu)?;

        // Use a filter node to add general predicates + timestamp
        // range, if any
        if let Some(predicate) = predicate {
            if let Some(filter_expr) = predicate.filter_expr() {
                // Rewrite expression so it only refers to columns in this chunk
                let schema = provider.iox_schema();
                trace!(%table_name, ?filter_expr, "Adding filter expr");
                let mut rewriter = MissingColumnsToNull::new(schema);
                let filter_expr =
                    filter_expr
                        .rewrite(&mut rewriter)
                        .context(RewritingFilterPredicateSnafu {
                            table_name: table_name.as_ref(),
                        })?;

                trace!(?filter_expr, "Rewritten filter_expr");

                plan_builder = plan_builder
                    .filter(filter_expr)
                    .context(BuildingPlanSnafu)?;
            }
        }

        Ok(ScanPlan {
            plan_builder,
            provider,
        })
    }
}
