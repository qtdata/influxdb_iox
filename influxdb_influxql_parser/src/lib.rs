//! # Parse a subset of [InfluxQL]
//!
//! [InfluxQL]: https://docs.influxdata.com/influxdb/v1.8/query_language

#![deny(rustdoc::broken_intra_doc_links, rust_2018_idioms)]
#![warn(
    missing_copy_implementations,
    missing_docs,
    clippy::explicit_iter_loop,
    clippy::future_not_send,
    clippy::use_self,
    clippy::clone_on_ref_ptr,
    clippy::todo,
    clippy::dbg_macro
)]

use crate::common::{statement_terminator, ws0};
use crate::internal::Error as InternalError;
use crate::statement::{statement, Statement};
use nom::combinator::eof;
use nom::Offset;
use std::fmt::{Debug, Display, Formatter};

#[cfg(test)]
mod test_util;

pub mod common;
pub mod create;
pub mod delete;
pub mod drop;
pub mod explain;
pub mod expression;
pub mod identifier;
mod internal;
mod keywords;
pub mod literal;
pub mod parameter;
pub mod select;
pub mod show;
pub mod show_field_keys;
pub mod show_measurements;
pub mod show_retention_policies;
pub mod show_tag_keys;
pub mod show_tag_values;
pub mod simple_from_clause;
pub mod statement;
pub mod string;
pub mod visit;
pub mod visit_mut;

/// A error returned when parsing an InfluxQL query using
/// [`parse_statements`] fails.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParseError {
    message: String,
    pos: usize,
}

impl Display for ParseError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} at pos {}", self.message, self.pos)
    }
}

/// ParseResult is type that represents the success or failure of parsing
/// a given input into a set of InfluxQL statements.
///
/// Errors are human-readable messages indicating the cause of the parse failure.
pub type ParseResult = Result<Vec<Statement>, ParseError>;

/// Parse the input into a set of InfluxQL statements.
pub fn parse_statements(input: &str) -> ParseResult {
    let mut res = Vec::new();
    let mut i: &str = input;

    loop {
        // Consume whitespace from the input
        i = match ws0(i) {
            Ok((i1, _)) => i1,
            _ => unreachable!("ws0 is infallible"),
        };

        if eof::<_, nom::error::Error<_>>(i).is_ok() {
            return Ok(res);
        }

        if let Ok((i1, _)) = statement_terminator(i) {
            i = i1;
            continue;
        }

        match statement(i) {
            Ok((i1, o)) => {
                res.push(o);
                i = i1;
            }
            Err(nom::Err::Failure(InternalError::Syntax {
                input: pos,
                message,
            })) => {
                return Err(ParseError {
                    message: message.into(),
                    pos: input.offset(pos),
                })
            }
            // any other error indicates an invalid statement
            Err(_) => {
                return Err(ParseError {
                    message: "invalid SQL statement".into(),
                    pos: input.offset(i),
                })
            }
        }
    }
}

#[cfg(test)]
mod test {
    use crate::parse_statements;

    /// Validates that the [`parse_statements`] function
    /// handles statement terminators and errors.
    #[test]
    fn test_parse_statements() {
        // Parse a single statement, without a terminator
        let got = parse_statements("SHOW MEASUREMENTS").unwrap();
        assert_eq!(format!("{}", got.first().unwrap()), "SHOW MEASUREMENTS");

        // Parse a single statement, with a terminator
        let got = parse_statements("SHOW MEASUREMENTS;").unwrap();
        assert_eq!(format!("{}", got[0]), "SHOW MEASUREMENTS");

        // Parse multiple statements with whitespace
        let got = parse_statements("SHOW MEASUREMENTS;\nSHOW MEASUREMENTS LIMIT 1").unwrap();
        assert_eq!(format!("{}", got[0]), "SHOW MEASUREMENTS");
        assert_eq!(format!("{}", got[1]), "SHOW MEASUREMENTS LIMIT 1");

        // Parse multiple statements with a terminator in quotes, ensuring it is not interpreted as
        // a terminator
        let got =
            parse_statements("SHOW MEASUREMENTS WITH MEASUREMENT = \";\";SHOW DATABASES").unwrap();
        assert_eq!(
            format!("{}", got[0]),
            "SHOW MEASUREMENTS WITH MEASUREMENT = \";\""
        );
        assert_eq!(format!("{}", got[1]), "SHOW DATABASES");

        // Parses a statement with a comment
        let got = parse_statements(
            "SELECT idle FROM cpu WHERE host = 'host1' --GROUP BY host fill(null)",
        )
        .unwrap();
        assert_eq!(
            format!("{}", got[0]),
            "SELECT idle FROM cpu WHERE host = 'host1'"
        );

        // Parses multiple statements with a comment
        let got = parse_statements(
            "SELECT idle FROM cpu WHERE host = 'host1' --GROUP BY host fill(null)\nSHOW DATABASES",
        )
        .unwrap();
        assert_eq!(
            format!("{}", got[0]),
            "SELECT idle FROM cpu WHERE host = 'host1'"
        );
        assert_eq!(format!("{}", got[1]), "SHOW DATABASES");

        // Parses statement with inline comment
        let got = parse_statements(r#"SELECT idle FROM cpu WHERE/* time > now() AND */host = 'host1' --GROUP BY host fill(null)"#).unwrap();
        assert_eq!(
            format!("{}", got[0]),
            "SELECT idle FROM cpu WHERE host = 'host1'"
        );

        // Returns error for invalid statement
        let got = parse_statements("BAD SQL").unwrap_err();
        assert_eq!(format!("{}", got), "invalid SQL statement at pos 0");

        // Returns error for invalid statement after first
        let got = parse_statements("SHOW MEASUREMENTS;BAD SQL").unwrap_err();
        assert_eq!(format!("{}", got), "invalid SQL statement at pos 18");
    }
}
