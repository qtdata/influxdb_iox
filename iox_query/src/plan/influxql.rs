use crate::{DataFusionError, IOxSessionContext, QueryNamespace};
use datafusion::common::Result;
use datafusion::execution::context::SessionState;
use datafusion::logical_expr::LogicalPlan;
use influxdb_influxql_parser::statement::Statement;
use std::sync::Arc;

/// InfluxQL query planner
#[allow(unused)]
#[derive(Debug)]
pub struct InfluxQLToLogicalPlan<'a> {
    ctx: &'a IOxSessionContext,
    state: SessionState,
    database: Arc<dyn QueryNamespace>,
}

impl<'a> InfluxQLToLogicalPlan<'a> {
    pub fn new(ctx: &'a IOxSessionContext, database: Arc<dyn QueryNamespace>) -> Self {
        Self {
            ctx,
            state: ctx.inner().state(),
            database,
        }
    }

    pub fn statement_to_plan(&self, statement: Statement) -> Result<LogicalPlan> {
        match statement {
            Statement::CreateDatabase(_) => {
                Err(DataFusionError::NotImplemented("CREATE DATABASE".into()))
            }
            Statement::Delete(_) => Err(DataFusionError::NotImplemented("DELETE".into())),
            Statement::DropMeasurement(_) => {
                Err(DataFusionError::NotImplemented("DROP MEASUREMENT".into()))
            }
            Statement::Explain(_) => Err(DataFusionError::NotImplemented("EXPLAIN".into())),
            Statement::Select(_) => Err(DataFusionError::NotImplemented("SELECT".into())),
            Statement::ShowDatabases(_) => {
                Err(DataFusionError::NotImplemented("SHOW DATABASES".into()))
            }
            Statement::ShowMeasurements(_) => {
                Err(DataFusionError::NotImplemented("SHOW MEASUREMENTS".into()))
            }
            Statement::ShowRetentionPolicies(_) => Err(DataFusionError::NotImplemented(
                "SHOW RETENTION POLICIES".into(),
            )),
            Statement::ShowTagKeys(_) => {
                Err(DataFusionError::NotImplemented("SHOW TAG KEYS".into()))
            }
            Statement::ShowTagValues(_) => {
                Err(DataFusionError::NotImplemented("SHOW TAG VALUES".into()))
            }
            Statement::ShowFieldKeys(_) => {
                Err(DataFusionError::NotImplemented("SHOW FIELD KEYS".into()))
            }
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::exec::{Executor, ExecutorType};
    use crate::test::{TestChunk, TestDatabase};
    use assert_matches::assert_matches;
    use influxdb_influxql_parser::parse_statements;
    use predicate::rpc_predicate::QueryNamespaceMeta;

    fn logical_plan(sql: &str) -> Result<LogicalPlan> {
        let mut statements = parse_statements(sql).unwrap();
        // setup a db
        let chunk0 = Arc::new(
            TestChunk::new("h2o")
                .with_id(0)
                .with_tag_column("foo")
                .with_tag_column("bar")
                .with_i64_field_column("i64_field")
                .with_i64_field_column("i64_field_2")
                .with_time_column()
                .with_one_row_of_data(),
        );
        // index of columns in the above chunk: [bar, foo, i64_field, i64_field_2, time]
        let executor = Arc::new(Executor::new_testing());
        let ctx = executor.new_context(ExecutorType::Query);
        let test_db = Arc::new(TestDatabase::new(Arc::clone(&executor)));
        test_db.add_chunk("my_partition_key", Arc::clone(&chunk0));
        let table = "h2o";

        let _schema = test_db.table_schema(table).unwrap();

        let planner = InfluxQLToLogicalPlan::new(&ctx, test_db);
        planner.statement_to_plan(statements.pop().unwrap())
    }

    /// Verify the list of unsupported statements.
    ///
    /// Some statements will remain unsupported, indefinitely.
    #[test]
    fn test_unsupported_statements() {
        assert_matches!(logical_plan("CREATE DATABASE foo"), Err(DataFusionError::NotImplemented(s)) => assert_eq!(s, "CREATE DATABASE"));
        assert_matches!(logical_plan("DELETE FROM foo"), Err(DataFusionError::NotImplemented(s)) => assert_eq!(s, "DELETE"));
        assert_matches!(logical_plan("DROP MEASUREMENT foo"), Err(DataFusionError::NotImplemented(s)) => assert_eq!(s, "DROP MEASUREMENT"));
        assert_matches!(logical_plan("EXPLAIN SELECT bar FROM foo"), Err(DataFusionError::NotImplemented(s)) => assert_eq!(s, "EXPLAIN"));
        assert_matches!(logical_plan("SELECT bar FROM foo"), Err(DataFusionError::NotImplemented(s)) => assert_eq!(s, "SELECT"));
        assert_matches!(logical_plan("SHOW DATABASES"), Err(DataFusionError::NotImplemented(s)) => assert_eq!(s, "SHOW DATABASES"));
        assert_matches!(logical_plan("SHOW MEASUREMENTS"), Err(DataFusionError::NotImplemented(s)) => assert_eq!(s, "SHOW MEASUREMENTS"));
        assert_matches!(logical_plan("SHOW RETENTION POLICIES"), Err(DataFusionError::NotImplemented(s)) => assert_eq!(s, "SHOW RETENTION POLICIES"));
        assert_matches!(logical_plan("SHOW TAG KEYS"), Err(DataFusionError::NotImplemented(s)) => assert_eq!(s, "SHOW TAG KEYS"));
        assert_matches!(logical_plan("SHOW TAG VALUES WITH KEY = bar"), Err(DataFusionError::NotImplemented(s)) => assert_eq!(s, "SHOW TAG VALUES"));
        assert_matches!(logical_plan("SHOW FIELD KEYS"), Err(DataFusionError::NotImplemented(s)) => assert_eq!(s, "SHOW FIELD KEYS"));
    }
}
