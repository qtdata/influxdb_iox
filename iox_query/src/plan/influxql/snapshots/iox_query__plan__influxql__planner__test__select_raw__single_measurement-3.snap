---
source: iox_query/src/plan/influxql.rs
expression: "plan(\"SELECT time as timestamp, f64_field FROM data\")"
---
Projection: data.time AS timestamp, data.f64_field AS f64_field [timestamp:Timestamp(Nanosecond, None), f64_field:Float64;N]
  TableScan: data [TIME:Boolean;N, bar:Dictionary(Int32, Utf8);N, bool_field:Boolean;N, f64_field:Float64;N, foo:Dictionary(Int32, Utf8);N, i64_field:Int64;N, mixedCase:Float64;N, str_field:Utf8;N, time:Timestamp(Nanosecond, None), with space:Float64;N]
